drivers/sensor/bme680/bme680.c

```diff
@@ -0,0 +1,444 @@
+/* bme680.c - Driver for Bosch Sensortec's BME680 temperature, pressure,
+ * humidity and gas sensor
+ */
+
+/*
+ * Copyright (c) 2016, 2017 Intel Corporation
+ * Copyright (c) 2017 IpTronix S.r.l.
+ * Copyright (c) 2018 Bosch Sensortec GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <sensor.h>
+#include <init.h>
+#include <gpio.h>
+#include <misc/byteorder.h>
+#include <misc/__assert.h>
+
+#include <i2c.h>
+
+#include "bme680.h"
+
+static int bme680_reg_read(struct bme680_data *data, u8_t start, u8_t *buf,
+			   int size)
+{
+	return i2c_burst_read(data->i2c_master, data->i2c_slave_addr, start,
+			      buf, size);
+	return 0;
+}
+
+static int bme680_reg_write(struct bme680_data *data, u8_t reg, u8_t val)
+{
+	return i2c_reg_write_byte(data->i2c_master, data->i2c_slave_addr,
+				  reg, val);
+	return 0;
+}
+
+static void bme680_calc_temp(struct bme680_data *data, u32_t adc_temp)
+{
+	s64_t var1;
+	s64_t var2;
+	s64_t var3;
+
+	var1 = ((s32_t)adc_temp >> 3) - ((s32_t)data->par_t1 << 1);
+	var2 = (var1 * (s32_t)data->par_t2) >> 11;
+	var3 = ((var1 >> 1) * (var1 >> 1)) >> 12;
+	var3 = ((var3) * ((s32_t)data->par_t3 << 4)) >> 14;
+	data->t_fine = (s32_t)(var2 + var3);
```

Comment:
From our experience, different compilers, especially 8 and 16 bit targets, compile the code differently and this results in overflows and saturation and hence, taking this into account, the explicit castings have been added to prevent these cases in the field.